[ #6 ]로그인 구현 및 테스트코드 작성

gr2nsky opened the pull request:
### Overview
1. SpringSecurity - JWT 기반 로그인 인증 구축
2. 로그인 API 작성
3. 로그인 관련 테스트코드 작성

* 해당 PR merge 이후 이전 PR(회원가입 및 테스트코드 작성) 피드백 사항을 refac 라벨을 달아 추가로 PR 드리려고 합니다.

gr2nsky committed:
[ #3 #5 ] SpringSecurity - JWT 구축

1. User 테이블에 Role 속성 추가 (USER / ADMIN)
  - 비록 관리자 페이지를 작성할지는 미지수이나, 서비스 관리를 위해선 꼭 필요하므로 추가
2. JWT 구현
  - 관련 class들은 rootDir/global/jwt 디렉토리에 모아 두었음
3. SpringSecurity 수정
  - jwt 인증 filter 추가
  - ExceptionHandling 추가
  - authorizeHttpRequests 수정

gr2nsky committed:
[ #5] 로그인 API 구현

로그인 API 추가 및 관련 request/response, exception 추가

gr2nsky committed:
[ #1 #5 ]로그인, 회원가입, jwt관련 버그 수정

1. 오탈자 및 몇몇 표현 수정
2. SecurityConfig에서 authorizeHttpRequests의 requestMatchers에 permitAll로 지정한 URL이 유효하지 않아 수정
3. UserDetailService에 누락된 Serivce 어노테이션 추가
4.jjwt 라이브러리(버전 0.9.1)가 Java 11 이상에서 제거된 javax.xml.bind 패키지를 참조하려고 하나, java11에서는 해당 클래스가 제거되어서 오류가 발생. javax.xml.bind:jaxb-api 종속성을 빌드 구성에 추가하여 해결
5. 회원가입시 암호화된 비밀번호를 암호화된 전화번호로 덮어씌우는 소스때문에 올바른 비밀번호를 입력해도 로그인이 되지않던 오류 수정
, 더불어 joinRequest객체에 setMdn 메서드 추가

gr2nsky committed:
[ #5 ] 로그인 관련 테스트코드 작성 및 로그인 일부 수정

1. LoginRequest 객체 생성자에도 builder를 사용해 다른 코드들과 통일감 유지
2. UserService에서 email로 사용자를 찾는 select query 수행부분을 분리해 테스트에 사용
3. 전반적인 Test code들에서 객체생성시 builder를 간헐적으로 사용하던 것을 모두 builder로 객체를 생성하도록 수정
4. UserServiceTest와 UserControllerTest에 Login관련 테스트코드 추가

f-lab-kyle commented on src/main/java/com/flab/urlumberjack/global/config/SecurityConfig.java:
```diff
@@ -4,24 +4,55 @@
-			.formLogin((formLogin) -> formLogin.disable())
-			.httpBasic((httpBasic) -> httpBasic.disable())
+			//basic setting
+			.csrf(AbstractHttpConfigurer::disable)
+			.formLogin(AbstractHttpConfigurer::disable)
+			.httpBasic(AbstractHttpConfigurer::disable)
```
훨씬 보기 좋네요 👍

f-lab-kyle commented on src/main/java/com/flab/urlumberjack/global/config/SecurityConfig.java:
```diff
@@ -4,24 +4,55 @@
 				authorizeHttpRequests
-					.requestMatchers("/join/**").permitAll()
-					.anyRequest().permitAll()
+					//현재 구체적인 기능 api들이 작성되지 않았으므로, 이후 비 로그인 사용자가 사용할 수 있는 api url을 지속적으로 추가해 주어야 함
+					.requestMatchers("/api/v1/user/join/**", "/api/v1/user/login/**", "/api/v1/main/**").permitAll()
```
고민하신것처럼, API가 추가적으로 발생할때마다 이 패턴을 조정해야 할지 개발 범위를 전체적으로 고민하면서 미리 정의해봐도 좋을것 같아요.

f-lab-kyle commented on src/main/java/com/flab/urlumberjack/global/jwt/JwtAccessDeniedHandler.java:
```diff
@@ -0,0 +1,30 @@
+	public void handle(HttpServletRequest request,
+		HttpServletResponse response,
+		AccessDeniedException accessDeniedException) throws IOException {
+		log.error("JWT - returns 403 forbidden. Message : {}", accessDeniedException.getMessage());
```
로깅레벨을 고민해봐도 좋을것 같습니다. 실 서비스라 가정하면 생각보다 꽤 많이 발생할 로그인것 같아서요. (to warn?)

f-lab-kyle commented on src/main/java/com/flab/urlumberjack/global/jwt/JwtProvider.java:
```diff
@@ -0,0 +1,118 @@
+	}
+
+	//@PostConstruct 어노테이션을 활용해, 의존성 주입이 이루어진 후에 secretKey를 Base64로 인코딩한다.
+	@PostConstruct
+	protected void init() {
+		secretKey = Base64.getEncoder().encodeToString(secretKey.getBytes());
+	}
```
추후 디버깅을 위해서라도 어플리케이션 초기 실행시 로깅을 해둬도 괜찮을것 같아요.

f-lab-kyle commented on src/main/java/com/flab/urlumberjack/global/jwt/JwtProvider.java:
```diff
@@ -0,0 +1,118 @@
+		Claims claims = Jwts.claims().setSubject(email);
+		claims.put(ROLE, role);
+
+		Date now = new Date();
+		Date validity = new Date(now.getTime() + exp);
+
```
시간 계산은 Date `+` 연산보다 다른 클래스 (e.g. LocalDateTime) 를 사용하는게 안전/명시적일것 같아요.

f-lab-kyle commented on src/main/java/com/flab/urlumberjack/global/jwt/JwtProvider.java:
```diff
@@ -0,0 +1,118 @@
+
+	private String secretKey;
+	private long exp;
+	private final UserDetailsService service;
```
```suggestion
	private final UserDetailsService userDetailsService;
```

f-lab-kyle commented on src/main/java/com/flab/urlumberjack/user/service/UserService.java:
```diff
@@ -8,22 +8,31 @@
 public class UserService {
 
-	final UserMapper mapper;
+	private UserMapper mapper;
```
이것만 final 을 제외한 이유가 있을까요?

f-lab-kyle commented on src/main/java/com/flab/urlumberjack/user/service/UserService.java:
```diff
@@ -8,22 +8,31 @@
-	final PasswordEncoder passwordEncoder;
-
-	public UserService(UserMapper mapper, PasswordEncoder passwordEncoder) {
+	public UserService(UserMapper mapper, PasswordEncoder passwordEncoder, JwtProvider jwtProvider) {
```
lombok을 활용해도 좋을것 같아요.

f-lab-kyle commented on src/main/java/com/flab/urlumberjack/user/service/UserService.java:
```diff
@@ -33,7 +42,7 @@ public JoinResponse join(JoinRequest joinRequest) {
 		String encryptMdn = passwordEncoder.encode(joinRequest.getMdn());
 		joinRequest.setPw(encryptPassword);
-		joinRequest.setPw(encryptMdn);
+		joinRequest.setMdn(encryptMdn);
```
👍

f-lab-kyle commented on src/main/java/com/flab/urlumberjack/user/service/UserService.java:
```diff
@@ -47,10 +56,34 @@ public JoinResponse join(JoinRequest joinRequest) {
+		User user = selectUserByEmail(loginRequest.getEmail());
+
+		if (!isMatchedPassword(loginRequest.getPw(), user.getPw())) {
+			throw new WrongPasswordException(ErrorMessage.WRONG_PASSWORD);
```
다른 예외도 마찬가지 인데 예외 : 메세지 = 1:1 라면 따로 에러 메세지를 클래스로 관리 안하고 예외 내부에 설정해도 괜찮을것 같긴 하네요. 다만 에러메세지 사용처가 다른곳에서도 있다면 무시해주시구요 ㅎ

f-lab-kyle commented on src/test/java/com/flab/urlumberjack/user/controller/UserControllerTest.java:
```diff
@@ -61,7 +66,11 @@ void when_allFieldsAreEntered_expect_joinToSuccess() throws Exception {
 	@ValueSource(strings = {" ", "abc123", "gamil.com", "123!naver.com"})
 	@DisplayName("email 필드가 null, empty, blank 상태이거나 email형식이 아니라면 회원가입이 실패한다.")
```
null에 대해선 테스트가 안되고 있는것 같아요! 의도한게 아니라면 displayName을 수정하면 될것 같아요.

f-lab-kyle commented on src/test/java/com/flab/urlumberjack/user/service/UserServiceTest.java:
```diff
@@ -5,25 +5,38 @@
 	UserMapper mapper;
 
 	@InjectMocks
+	UserService mcokService;
```
오타인것 같아요

f-lab-kyle commented on src/test/java/com/flab/urlumberjack/user/service/UserServiceTest.java:
```diff
@@ -5,25 +5,38 @@
+import com.flab.urlumberjack.user.exception.NotExistedUserException;
 import com.flab.urlumberjack.user.mapper.UserMapper;
 
 @ExtendWith(MockitoExtension.class)
+@SpringBootTest
 class UserServiceTest {
```
어떤건 mocking, 어떤건 bean을 띄워 테스트 하는걸로 보이는데 그렇다면 테스트 클래스를 분리해도 좋을것 같아요.

f-lab-kyle reviewed (commented):
수고하셨습니다~

gr2nsky committed:
[ #6 ] 로그인 feature 에 대한 코드리뷰 피드백 반영

1. SecurityConfig에서 비회원만 접근 가능한 URL을 명시하고 그 외엔 권한이 요구되도록 표시
  - 비회원 접근 API가 훨씬 적으므로, 비회원이 접근가능한 API 외 권한이 요구되도록 처리
2. 유효하지 않은 token 접근 로그레벨을 warn으로 변경
  - 프로그램보다는 사용자에게서 기인하는 오류이며, 발생빈도가 높을거라고 사료됨
3. JWT secret key 생성시 debug 레벨에서 로그를 남기도록 수정
  - 추후 디버깅을 위함
4. 토큰 생성시 날자 연산은 LocalDateTime으로 수행하도록 수정
  - Date는 가변객체라 수정이 가능해 의도치않게 동작할 수 있고, 스레드 안정성이 떨어저 동시에 여러 스레드가 접근할 경우 동기화처리를 해야 한다. 그리고 Deprecated된 메서드들이 존재해 향후 문제가 발생할 가능성이 크다.
5. custom exception 전반적인 수정
  - 이제 RuntimeException를 상속받던 클래스들은 CustomRuntimeException를 상속받아 사용합니다. CustomRuntimeException은 RuntimeException를 상속받으며, 필드로 HttpStatus타입의 변수를 가지고 있으며, 생성자로는 message와 HttpStatus가 필요합니다. 메서드는 getHttpStatus를 가지고 있습니다.
  - 내부에 public static String message 객체를 선언해야 합니다. CustomRuntimeException생성자 호출 및 외부 참조에 필요합니다,.
  - 생성자 생성시 httpStatusCode를 입력해야 합니다.
  - 예외 클래스들의 수정에 따라 ErrorResponse클래스의 글로벌 생성자도 수정이 되었습니다.
6. Exception에 대한 message들이 분리됨에 따라 ErrorMessage class는 에러에 관련한 httpStatus와 message를 갖게끔 분리되었습니다.
7. 파라미터를 테스트하는 테스트코드의 경우 입력받은 파라미터값 확인을 위해 로깅 추가
8. userServiceTest에서 mock객체를 사용하는 테스트와 bean 객체를 사용하는 테스트를 별도 클래스로 분리

gr2nsky committed:
[ #5 ] Login feature 피드백 반영

1. UserController의 Join method mapping 수정
    - post는 join의 의미를 가지고 있으며, user를 등록하는 api는 가입 하나뿐이므로 mapping을 controller mapping을 그대록 사용하도록 수정
2. 구체적이지 않은 이름을 가진 User domain값과 클래스 수정
    - pw -> password
    - type -> UserJoinType
    - Status -> UserAccountStatus
3. insert 수행에 대한 성공/실패를 판별하는 global util 클래스 및 메서드 생성
    - 회원가입 수행시에 사용
4. isDuplicatedEmail 메서드명을 checkDuplicatedEmail로 수정
    - 조건 충족시, 반환값 없이 예외를 발생시키므로 좀더 알맞은 네이밍으로 수정
5. select 쿼리들 중, *를 사용하는 구문 수정
6. insertUser query에서 type값에 따른 choose구문 제거
    - joinReqeust의 기본값을 DEFAULT로 설정해 불필요한 choose구문 제거
7. TestClass들에 DisplayName 어노테이션을 추가
8. 이전 커밋에서 분리했던 UserServiceTest, UserServiceMockTest에서 사용하지않는 불필요한 어노테이션 정리

gr2nsky committed:
[ #6 ] refresh token - used spring cache

구현결과, 캐시 만료시 자동 삭제되는 기능은 스프링 캐시에서 지원하지 않아 이후 카페인 캐시로 구현해야 함. 작업분 저장을 위해 커밋

gr2nsky committed:
[ #6 ] refresh token - used caffeine cache

캐시 만료정책 설정을 위해 스프링 캐시 대신 카페인 캐시로 리프레시 토큰 을 구현

gr2nsky committed:
리프레시 토큰 테스트코드 작성

gr2nsky committed:
Merge branch 'feature/user-login' into feature/user-join

gr2nsky committed:
Merge pull request #7 from f-lab-edu/feature/user-join

[ #5 ] join feature 피드백 반영
